### PR TITLE
New AST nodes for f-string elements

### PR DESCRIPTION
**Pull request by dhruvmanila**
Rebase of #6365 authored by @davidszotten.

## Summary

This PR updates the AST structure for an f-string elements.

The main **motivation** behind this change is to have a dedicated node for the string part of an f-string. Previously, the existing `ExprStringLiteral` node was used for this purpose which isn't exactly correct. The `ExprStringLiteral` node should include the quotes as well in the range but the f-string literal element doesn't include the quote as it's a specific part within an f-string. For example,

```python
f"foo {x}"
# ^^^^
# This is the literal part of an f-string
```

The introduction of `FStringElement` enum is helpful which represent either the literal part or the expression part of an f-string.

### Rule Updates

This means that there'll be two nodes representing a string depending on the context. One for a normal string literal while the other is a string literal within an f-string. The AST checker is updated to accommodate this change. The rules which work on string literal are updated to check on the literal part of f-string as well.

#### Notes

1. The `Expr::is_literal_expr` method would check for `ExprStringLiteral` and return true if so. But now that we don't represent the literal part of an f-string using that node, this improves the method's behavior and confines to the actual expression. We do have the `FStringElement::is_literal` method.
2. We avoid checking if we're in a f-string context before adding to `string_type_definitions` because the f-string literal is now a dedicated node and not part of `Expr`.
3. Annotations cannot use f-string so we avoid changing any rules which work on annotation and checks for `ExprStringLiteral`.

## Test Plan

- All references of `Expr::StringLiteral` were checked to see if any of the rules require updating to account for the f-string literal element node.
- New test cases are added for rules which check against the literal part of an f-string.
- Check the ecosystem results and ensure it remains unchanged.

## Performance

There's a performance penalty in the parser. The reason for this remains unknown as it seems that the generated assembly code is now different for the `__reduce154` function. The reduce function body is just popping the `ParenthesizedExpr` on top of the stack and pushing it with the new location.

- The size of `FStringElement` enum is the same as `Expr` which is what it replaces in `FString::format_spec`
- The size of `FStringExpressionElement` is the same as `ExprFormattedValue` which is what it replaces

I tried reducing the `Expr` enum from 80 bytes to 72 bytes but it hardly resulted in any performance gain. The difference can be seen here:
- Original profile: https://share.firefox.dev/3Taa7ES
- Profile after boxing some node fields: https://share.firefox.dev/3GsNXpD

### Backtracking

I tried backtracking the changes to see if any of the isolated change produced this regression. The problem here is that the overall change is so small that there's only a single checkpoint where I can backtrack and that checkpoint results in the same regression. This checkpoint is to revert using `Expr` to the `FString::format_spec` field. After this point, the change would revert back to the original implementation.

## Review process

The review process is similar to #7927. The first set of commits update the node structure, parser, and related AST files. Then, further commits update the linter and formatter part to account for the AST change.

### CODE REVIEW

**Comment by dhruvmanila** [comment]
Current dependencies on/for this PR:
* main
  * **PR #8062** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/8062?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #8063** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/8063?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #8064** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/8064?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
        * **PR #7927** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/7927?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
          * **PR #8826** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/8826?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
            * **PR #8828** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/8828?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
              * **PR #8835** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/8835?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/astral-sh/ruff/8835?utm_source=stack-comment).

**Comment by dhruvmanila** [comment]
Ugh, I need to revisit this later. I think I messed up the git history somehow :/

File: crates/ruff_python_ast/src/nodes.rs [hunk 1]
```diff
@@ -179,7 +179,7 @@ pub struct StmtFunctionDef {
     pub is_async: bool,
     pub decorator_list: Vec<Decorator>,
     pub name: Identifier,
-    pub type_params: Option<TypeParams>,
+    pub type_params: Option<Box<TypeParams>>,
```
Comment (MichaReiser) [inline crates/ruff_python_ast/src/nodes.rs:182]: What's the motivation for boxing some fields? it seems unrelated to the PR itself.

File: crates/ruff_python_ast/src/nodes.rs [hunk 2]
```diff
@@ -3934,11 +3939,32 @@ mod size_assertions {
...
+    assert_eq_size!(Stmt, [u8; 120]);
+    assert_eq_size!(StmtFunctionDef, [u8; 120]);
     assert_eq_size!(StmtClassDef, [u8; 104]);
     assert_eq_size!(StmtTry, [u8; 112]);
     assert_eq_size!(Expr, [u8; 80]);
     assert_eq_size!(Pattern, [u8; 96]);
     assert_eq_size!(Mod, [u8; 32]);
 }
+
+impl Ranged for crate::nodes::FStringExpressionElement {
```
Comment (MichaReiser) [inline crates/ruff_python_ast/src/nodes.rs:3951]: Nit: Can we move this above the size_assertions?

File: crates/ruff_linter/src/rules/flake8_pytest_style/rules/helpers.rs [hunk 3]
```diff
@@ -58,15 +58,25 @@ pub(super) fn is_empty_or_null_string(expr: &Expr) -> bool {
         Expr::FString(ast::ExprFString { value, .. }) => {
             value.parts().all(|f_string_part| match f_string_part {
```
Comment (MichaReiser) [inline crates/ruff_linter/src/rules/flake8_pytest_style/rules/helpers.rs:59]: I got rather confused by the `parts` method here. Like what's the difference between a `FStringPart` and an `FStringElement`, they even have the same members (although `FStringPart::Literal` takes a `StringLiteral` and `Element::Literal` an `FStringElementLiteral`). 

It clicked when opening the PR and looking up the definition, although I wonder if we can improve the naming (unrelated to this PR). 

A not fully formed idea:

* `ExprStringLiteral` -> `ExprString`
* `ExprStringLiteral::parts` rename to `literals` (it returns the individal string literals of which the string expression is composed)
* `ExprFString::parts` rename to `literals`
* `FStringPart` rename to `ExprFStringLiteral` 
* `FStringPart::Literal` rename to `FSTringPart::String`

Same for bytes. 

We may also want to change the `parts` and `parts_mut` methods to return a `&[]` instead of an iterator. 

The part I feel uncertain about is if it is okay to name an `f"abcd"` a literal, but I think it is. For example, the object notation `{a: "bcd"}` in JavaScript is a `ObjectLiteral` and it contains dynamic content as well. 

I don't suggest making these changes as part of this PR. It's just something I dripped over when reviewing and thought I would raise.


Edit: Even the Python spec names them literals: A formatted string literal

File: crates/ruff_linter/src/rules/pyflakes/rules/f_string_missing_placeholders.rs [hunk 4]
```diff
@@ -47,11 +47,12 @@ impl AlwaysFixableViolation for FStringMissingPlaceholders {
...
-        .value
-        .f_strings()
-        .any(|f_string| f_string.values.iter().any(Expr::is_formatted_value_expr))
-    {
+    if expr.value.f_strings().any(|f_string| {
+        f_string
+            .elements
+            .iter()
+            .any(ast::FStringElement::is_expression)
+    }) {
```
Comment (MichaReiser) [inline crates/ruff_linter/src/rules/pyflakes/rules/f_string_missing_placeholders.rs:55]: Nit: We could add these as a method on `fstring` itself: `has_placeholders()` or similar if we want.

File: crates/ruff_linter/src/rules/tryceratops/rules/raise_vanilla_args.rs [hunk 5]
```diff
@@ -98,9 +98,13 @@ fn contains_message(expr: &Expr) -> bool {
...
-                                return true;
+                        for element in &f_string.elements {
+                            if let Some(ast::FStringLiteralElement { value, .. }) =
+                                element.as_literal()
+                            {
+                                if value.chars().any(char::is_whitespace) {
+                                    return true;
+                                }
                             }
                         }
```
Comment (MichaReiser) [inline crates/ruff_linter/src/rules/tryceratops/rules/raise_vanilla_args.rs:109]: Nit:
```suggestion
                    ast::FStringPart::FString(f_string) => {
                        for literal in f_string.elements.iter().filter_map(|element| element.as_literal()) {
                            if literal.value.chars().any(char::is_whitespace) {
                                return true;
                            }
                        }
```

File: crates/ruff_python_ast/src/visitor/transformer.rs [hunk 6]
```diff
@@ -743,6 +725,31 @@ pub fn walk_pattern_keyword<V: Transformer + ?Sized>(
...
+    for element in &mut f_string.elements {
+        visitor.visit_f_string_element(element);
+    }
+}
+
+pub fn walk_f_string_element<V: Transformer + ?Sized>(
+    visitor: &V,
+    f_string_element: &mut FStringElement,
+) {
+    if let ast::FStringElement::Expression(ast::FStringExpressionElement {
```
Comment (MichaReiser) [inline crates/ruff_python_ast/src/visitor/transformer.rs:738]: It feels inconsistent to me that we have visitors for `string_literal` and `byte_literals` but that we aren't visiting `Literal` elements here. I think we should either have visitors for both or neither. The same applies to `preorder` and the regular visitor

File: crates/ruff_python_ast/src/visitor/preorder.rs [hunk 7]
```diff
@@ -519,6 +518,29 @@ where
...
+            ..
+        }) = f_string_element
+        {
+            visitor.visit_expr(expression);
+            if let Some(format_spec) = format_spec {
+                for spec_element in &format_spec.elements {
+                    visitor.visit_f_string_element(spec_element);
+                }
+            }
+        }
```
Comment (MichaReiser) [inline crates/ruff_python_ast/src/visitor/preorder.rs:539]: I think you can call here into `element.visit_preorder(visitor)` considering that `FStringExpressionElement` implements `AstNode`

File: crates/ruff_python_ast/tests/preorder.rs [hunk 8]
```diff
@@ -225,12 +225,6 @@ impl PreorderVisitor<'_> for RecordVisitor {
         self.exit_node();
     }
```
Comment (MichaReiser) [inline crates/ruff_python_ast/tests/preorder.rs:226]: Can you add overrides for the new visitor functions or rewrite the test to use `enter_node` and `exit_node`?

File: crates/ruff_python_formatter/src/other/f_string_expression_element.rs [hunk 9]
```diff
@@ -0,0 +1,16 @@
+use ruff_python_ast::FStringExpressionElement;
+
+use crate::prelude::*;
+
+#[derive(Default)]
+pub struct FormatFStringExpressionElement;
+
+impl FormatNodeRule<FStringExpressionElement> for FormatFStringExpressionElement {
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/src/other/f_string_expression_element.rs:8]: I'm not sure if it's worth adding these implementations considering that they won't work. I rather have `fstring_expression_element.format()` to not compile than fail at runtime. 

Should we exclude them from the generate script. We can still add them when they become useful.

**Review by MichaReiser — APPROVED** [review]
Thanks for driving this long standing PR forward. I like what I see in the parser diffs. 

The majority of comments are nits, but there's the question whather we should visit the `FStringLiteralElement` in visitors and if some lint rules should check if it si a single element fstring to avoid false positives. 

I would prefer for @charliermarsh to also review the linter changes, considering that he has more context than I.

File: crates/ruff_linter/src/checkers/ast/analyze/expression.rs [hunk 10]
```diff
@@ -1006,6 +1007,27 @@ pub(crate) fn expression(expr: &Expr, checker: &mut Checker) {
...
                 }
             }
+            for element in value.elements() {
+                let Some(literal_element @ ast::FStringLiteralElement { value, range }) =
+                    element.as_literal()
+                else {
+                    continue;
+                };
+                if checker.enabled(Rule::HardcodedBindAllInterfaces) {
+                    flake8_bandit::rules::hardcoded_bind_all_interfaces(checker, value, *range);
```
[the remark this answers is not in the document]
Comment (dhruvmanila) [reply crates/ruff_linter/src/checkers/ast/analyze/expression.rs:1017]: It always triggered for such cases, we're just maintaining the behavior. The reason is that it could be useful for cases like `f"0.0.0.0{port}/path"`: https://play.ruff.rs/7b6e80ea-5858-4983-877c-1669a5cd538d

File: crates/ruff_linter/src/rules/flake8_pytest_style/rules/helpers.rs (hunk 3 above)
Comment (dhruvmanila) [reply crates/ruff_linter/src/rules/flake8_pytest_style/rules/helpers.rs:59]: Thanks! I'll explore this in a follow-up PR.

File: crates/ruff_linter/src/rules/pyflakes/rules/f_string_missing_placeholders.rs (hunk 4 above)
Comment (dhruvmanila) [reply crates/ruff_linter/src/rules/pyflakes/rules/f_string_missing_placeholders.rs:55]: I thought so but currently it's only being used in a single place.

File: crates/ruff_python_ast/src/visitor/transformer.rs (hunk 6 above)
Comment (dhruvmanila) [reply crates/ruff_python_ast/src/visitor/transformer.rs:738]: I see your point. Wouldn't `visit_f_string_element` cover visiting the literal element of an f-string? This is similar to what we have with `visit_match_case` and `visit_pattern`.

Currently, we've a visitor for an `FStringElement` which is what you're highlighting here. I'd prefer to keep the literal visitors because they help in the formatter comment placements. Is what you're suggesting to replace `visit_f_string_element` with `visit_f_string_literal_element` and `visit_f_string_expression_element`? So, the hierarchy would be:

```bash
FString # visit_f_string
 |- FStringLiteralElement # visit_f_string_literal_element
 |- FStringExpressionElement # visit_f_string_expression_element
```

File: crates/ruff_linter/src/rules/flake8_pyi/rules/string_or_bytes_too_long.rs [hunk 11]
```diff
@@ -44,25 +43,30 @@ impl AlwaysFixableViolation for StringOrBytesTooLong {
 }
 
 /// PYI053
-pub(crate) fn string_or_bytes_too_long(checker: &mut Checker, expr: &Expr) {
+pub(crate) fn string_or_bytes_too_long(checker: &mut Checker, node: AnyNodeRef) {
```
[the remark this answers is not in the document]
Comment (dhruvmanila) [reply crates/ruff_linter/src/rules/flake8_pyi/rules/string_or_bytes_too_long.rs:46]: Yeah, I think this might be a good idea (https://github.com/astral-sh/ruff/pull/9016).

File: crates/ruff_python_ast/src/visitor/transformer.rs (hunk 6 above)
Comment (MichaReiser) [reply crates/ruff_python_ast/src/visitor/transformer.rs:738]: > I see your point. Wouldn't visit_f_string_element cover visiting the literal element of an f-string? This is similar to what we have with visit_match_case and visit_pattern.

That makes sense. Thank's for explaining.

File: crates/ruff_linter/src/checkers/ast/analyze/expression.rs [hunk 12]
```diff
@@ -1006,6 +1007,27 @@ pub(crate) fn expression(expr: &Expr, checker: &mut Checker) {
...
+                }
```
[the remark this answers is not in the document]
Comment (dhruvmanila) [reply crates/ruff_linter/src/checkers/ast/analyze/expression.rs:1017]: Refer: https://github.com/astral-sh/ruff/pull/9016

**Review by charliermarsh — APPROVED** [review]
This looks great.

File: crates/ruff_linter/src/checkers/ast/analyze/expression.rs [hunk 13]
```diff
@@ -1006,6 +1007,30 @@ pub(crate) fn expression(expr: &Expr, checker: &mut Checker) {
                     pyupgrade::rules::unicode_kind_prefix(checker, string_literal);
                 }
             }
+            for literal in value.elements().filter_map(|element| element.as_literal()) {
```
[the remark this answers is not in the document]
Comment (dhruvmanila) [reply crates/ruff_linter/src/checkers/ast/analyze/expression.rs:1010]: This is a good point and yes it will be a false negative now. It's not just `f"/tmp" f"/bad"` but also `"/tmp" f"/bad"` (string and f-string)

We can introduce a method which iterates over the concatenated literal parts of an f-string. For example,

```python
"foo" f"bar {x} baz" "end"
```

The above code would return two strings: `"foobar "` and `" bazend"`. Here, I don't see a way to avoid string allocation here.

[the remark this answers is not in the document]
Comment (dhruvmanila) [reply crates/ruff_linter/src/checkers/ast/analyze/expression.rs:1010]: I'm moving ahead with this limitation for now but we can revisit if it proves to be a problem. My hunch is this shouldn't be but you never know 🤷‍♂️
